src/libLookup/Lookup.cpp

```diff
@@ -1252,88 +1323,201 @@ bool Lookup::ProcessGetDSInfoFromSeed(const bytes& message, unsigned int offset,
   return true;
 }
 
-void Lookup::SendMessageToRandomL2lDataProvider(const bytes& message) const {
+void Lookup::SendMessageToRandomL2lDataProvider(const bytes& message) {
   LOG_MARKER();
 
   if (message.empty()) {
     LOG_GENERAL(WARNING, "Ignoring sending empty message");
     return;
   }
 
-  lock_guard<mutex> lock(m_mutexL2lDataProviders);
-  if (0 == m_l2lDataProviders.size()) {
-    LOG_GENERAL(WARNING, "l2l data providers are empty");
-    return;
-  }
+  auto localL2lDataProvider = GetL2lDataProviders();
+  vector<uint128_t> ipFromLocal;
 
-  int index = RandomGenerator::GetRandomInt(m_l2lDataProviders.size());
-  auto resolved_ip = TryGettingResolvedIP(m_l2lDataProviders[index].second);
+  std::transform(localL2lDataProvider.begin(), localL2lDataProvider.end(),
+                 std::back_inserter(ipFromLocal), [this](const auto& node) {
+                   return TryGettingResolvedIP(node.second);
+                 });
 
-  Blacklist::GetInstance().Whitelist(
-      resolved_ip);  // exclude this l2lookup ip from blacklisting
-  Peer tmpPeer(resolved_ip,
-               m_l2lDataProviders[index].second.GetListenPortHost());
-  LOG_GENERAL(INFO, "Sending message to l2l: " << tmpPeer);
+  Peer randomPeer;
+  auto port = ENABLE_SEED_TO_SEED_COMMUNICATION ? P2P_SEED_CONNECT_PORT
+                                                : DEFAULT_SEED_PORT;
+
+  vector<uint128_t> ipFromDns;
+
+  if (!QUERY_DNS_FOR_SEED ||
+      !ObtainIpListFromDns(ipFromDns, L2L_DATA_PROVIDERS_DNS)) {
+    if (ipFromLocal.empty()) {
+      LOG_GENERAL(WARNING, "No L2lDataProvider to send to");
+      return;
+    }
+
+    int index = RandomGenerator::GetRandomInt(ipFromLocal.size());
+    randomPeer = Peer{ipFromLocal[index], port};
+  } else {
+    // Find if list of L2lDataProvider from dns that intersects with local list
+    vector<uint128_t> intersectedList;
+    std::sort(ipFromLocal.begin(), ipFromLocal.end());
+    std::sort(ipFromDns.begin(), ipFromDns.end());
+    std::set_intersection(ipFromLocal.begin(), ipFromLocal.end(),
+                          ipFromDns.begin(), ipFromDns.end(),
+                          std::back_inserter(intersectedList));
+
+    if (intersectedList.empty()) {
+      // Not found in DNS, need to add one into our local list
+      int index = RandomGenerator::GetRandomInt(ipFromDns.size());
+      auto ip = ipFromDns[index];
+      auto ipStr = IPConverter::ToStrFromNumericalIP(ip);
+
+      randomPeer = Peer{ip, port};
+
+      // Need to insert into our local list for fetching verification
+      bytes pubkeyBytes;
+      if (!ObtainPubKeyFromUrl(pubkeyBytes, ipStr, L2L_DATA_PROVIDERS_DNS)) {
+        LOG_GENERAL(WARNING, "Pubkey cannot be found!");
```

Comment:
If the public key is not found for one node because of an error, should we should still try for other nodes(if the key is found) so that `localL2lDataProvider` is not empty

Reply:
Got it, will add this.